1. Read the City Cemetery data into a pandas DataFrame. **Hint:** see the [`read_excel` function](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html).

In [62]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

cc_df = pd.read_excel('../data/Historic_Nashville_City_Cemetery_Interments__1846-1979.xlsx')
cc_df.head()

Volume                           Name Burial Month  Month No.  Burial Day  \
0     6B                Bell, Nancy Ann          Aug        8.0         6.0   
1     6W  Drake, Mary Edith Smith, Mrs.          May        5.0        19.0   
2     6D                  Davis, Bessie          Sep        9.0        19.0   
3     6C            Creech, Emily Green          Aug        8.0         NaN   
4     6W            Creech, Emily Green          Aug        8.0         NaN   

   Burial Year Sex Race   Age             Residence Cause of Death/Burial  \
0       1979.0   F    W  10.0                   NaN                   NaN   
1       1979.0   F    W  67.0         Madison, Tenn                Cancer   
2       1979.0   F    W  90.0             Nashville               Old Age   
3       1978.0   F    W  91.0             Nashville                   NaN   
4       1978.0   F    W  91.0  Nashville, Tennessee                   NaN   

                Ave             Section/Lot  \
0               Oak      sec. 28 NW, lot 12   
1  Locust, Boyd Lot                  lot 10   
2            Poplar  McAllister, E25, lot 8   
3        sect.ion 9                      15   
4               NaN                     NaN   

                                             Remarks  
0                                   J. L. Hughes lot  
1  wife of Alex M. Drake, Madison Funeral Home Of...  
2                            daughter of Helen Davis  
3                                           Magnolia  
4                                          Malgnolia

In [63]:
cc_df.tail()

Volume                 Name Burial Month  Month No.  Burial Day  \
19740  4-1846  Kingsley, A., Capt.          Sep        9.0        28.0   
19741  4-1846               infant          Sep        9.0        28.0   
19742  4-1846       White, Lucinda          Sep        9.0        29.0   
19743  5-1881     Higgins, Rebecca          Apr        4.0        10.0   
19744  5-1881             Eves, R.          Apr        4.0        25.0   

       Burial Year Sex Race   Age Residence Cause of Death/Burial       Ave  \
19740       1846.0   M    W  67.0      City                Feever     Cedar   
19741       1846.0   M    W   NaN      City             Diarherea   Central   
19742       1846.0   F    W  25.0      City           Consumption   Central   
19743          NaN   F    W  68.0      City         Heart Disease    Cherry   
19744          NaN   M   W.  54.0      City         Heart Disease  Magnolia   

      Section/Lot                               Remarks  
19740         lot  superintendent of Nashville Cemetery  
19741         lot                   son of John Bateman  
19742         200        Fees to be paid by M. Jennings  
19743         lot                box paid, grave on lot  
19744         lot                box paid, grave on lot

2. **Most Common Causes of Death.**  
   a. Find the 10 most common (known) recorded causes of death.  
   

In [64]:
counts = cc_df['Cause of Death/Burial'].value_counts()
top_10_causes = counts.head(10)
top_10_causes

Cause of Death/Burial
Consumption     1767
Unknown         1260
Still Born       841
Cholrea          667
Not Known        639
Old Age          604
Cholera          575
Pneumonia        516
Flux             469
Complication     446
Name: count, dtype: int64

b. Clean up some of the misspellings. Use the [`replace` method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html) to clean up some of the causes of death. For example, replace "Cholrea" with "Cholera" and "Still Born" with "Stillborn".   
  

In [65]:
top_25_causes = counts.head(25)
top_25_causes

Cause of Death/Burial
Consumption       1767
Unknown           1260
Still Born         841
Cholrea            667
Not Known          639
Old Age            604
Cholera            575
Pneumonia          516
Flux               469
Complication       446
Cold               421
Teething           411
Typhoid Fever      310
Complicated        302
Brain Fever        286
Dropsy             284
Stil Born          278
Measles            236
Measels            224
Whooping Cough     205
Small Pox          196
Scarlot Fever      192
Infl of Bowels     156
Dropsey            151
Teethin            141
Name: count, dtype: int64

In [74]:
corrections = {
    'Cholrea': 'Cholera',
    'Still Born': 'Stillborn',
    'Stil Born': 'Stillborn',
    'Not Known': 'Unknown',
    'Teethin': 'Teething',
    'Measles': 'Measels',
    'Complicated': 'Complication',
    'Feever': 'Fever'}
cc_df = cc_df.replace(corrections)
counts = cc_df['Cause of Death/Burial'].value_counts()
top_25_causes = counts.head(25)
top_25_causes

Cause of Death/Burial
Unknown           1899
Consumption       1767
Cholera           1242
Stillborn         1204
Complication       748
Old Age            604
Teething           552
Pneumonia          516
Flux               469
Measels            460
Cold               421
Typhoid Fever      310
Brain Fever        286
Dropsy             284
Fever              275
Whooping Cough     205
Small Pox          196
Scarlot Fever      192
Infl of Bowels     156
Dropsey            151
Infl of Brain      137
Diptheria          129
Child Bed          124
Heart Disease      114
Croup              107
Name: count, dtype: int64

 **Bonus (Save For Last):** One method of detecting and correcting misspellings is fuzzy matching using metrics like [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance). Check out [thefuzz](https://github.com/seatgeek/thefuzz) a library that implements fuzzy matching and apply it to detect and correct misspellings.  
   c. After correcting the spellings, find the 10 most common causes of death and create a bar chart to display your findings.  


3. Create a line chart showing the number of burials per year.

In [75]:
#make a count of how many burials in each year
sns.lineplot(data=cc_df, x='Burial Year', y='count') #replace count with whatever the varriable is

ValueError: Could not interpret value `count` for parameter `y`

4. **Burials per Month and Year**  
    a. Use the [crosstab function](https://pandas.pydata.org/docs/reference/api/pandas.crosstab.html) to build a summary table showing the number of burials per year per month. You should have one row per year and one column for month.  

b. After creating this summary table, make sure that the year labels are integers (no decimal part) and that the columns are sorted correctly.

c. Filter your table to show only the years 1846 through 1860.  

**Bonus** Display your table so that highlights any months that had over 100 burials. Hint: Look at styling in pandas: https://pandas.pydata.org/pandas-docs/version/1.0/user_guide/style.html  

d. Go back to your original DataFrame and filter to the year and month that had the largest number of burials. What are the most common causes of death for that month?

5. **Burials per Decade**  
    a. Create a "Decade" column to show the decade in which a burial occurred.    

b. Examine the burials per decade. Look at the total number of burials and the proportion (aka percentage) of male burials and female burials.  


    c. Create a clustered bar chart to show how male and female burials have changed over time.


### Bonus Questions

6. **Burials per Age Group**  
    a. Add a column to the original dataset to classify each row to one of the following categories (0-18, 19-25, 26-40, 41-64, and 65+). (Hint: One way to do this is using the [`cut` function](https://pandas.pydata.org/docs/reference/api/pandas.cut.html)).  
    b. Create a "Year Group" dividing the records up based on burial year. Use the following groups: Before 1880, 1880-1900, 1901-1920, After 1920.  
    c. Use a grouped bar chart to show how the proportion of burials per age group has changed over time. How would you describe the trends in terms of proportion of burials per age group over time?


7. **Last Names**  
    a. Create a new column titled Last Name. Extract the last name from the Name column by subsetting to all characters to the left of the comma. (Hint: See the pandas [string split function](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.split.html). When you do this, make sure to consider cases where the name does not contain a comma.  
    b. What are the most common last names that appear?

8. The Civil War ended in April, 1865, with the last full year of fighting occurring in 1864.    
   a. Create a new DataFrame, burials_1864 by subsetting the full data to the rows for 1864 burials.  
   b. For any rows whose name contains the word "infant" (capitalized or lowercase), replace the "Age" value with 0.   
   c. Create a new column called age_category. If the individual was age 0 put "Infant". If the individual was not an infant but under age 18, put "Child". Everyone aged 18 and above should be listed as "Adult".  
   d.  Find the mean age for each age group.    
   e.  Find the last infant to be buried in 1864. Two ways this could be accomplished are sorting by both the month number and burial day or by creating a date column (for example, by using the [`to_datetime` function](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html) and then sorting by this date column.